### PR TITLE
Make Navigation fallback selector private

### LINKED ISSUE #50372: Make Navigation specific Core Data selectors Private
As suggested in the discussion below, let's make the new Navigation-specific selectors added to Core Data private APIs.

------

> I think the block should be able to tell core data "hey I got (somehow) this data which is the entity i want to work with" and core data will adopt that entity (which was not retrieved via a selector/resolver) and then handle edits, undo, updates

I don't think this is an Issue. What you describe is already working today. It's implemented in this PR and [the spin off PoC I created](https://github.com/WordPress/gutenberg/pull/50126). To be clear here is the code:

```js
if ( record ) {
         // This is dispatched to Core Data
	dispatch.receiveEntityRecords(
		'postType',
		'wp_navigation',
		record
	);
}
```

> The selector is block specific and the endpoint is also block specific today (nothing else uses that). It's an utility endpoint created for the navigation block to avoid doing all that on the client. I

This is the point of contention. I don't think it's a major problem to have this selector. 

@draganescu In the interests of moving forward would you be open to making it a private selector and then make it public only if/when we confirm that other areas will also require it (for example the `Navigation` section in the Site View sidebar). We get many benefits by using Core Data here and there are also already many examples of the same pattern in use within Core Data.

_Originally posted by @getdave in https://github.com/WordPress/gutenberg/pull/50032#discussion_r1180242261_

### PR DESCRIPTION
**Pull request by getdave**
## What?

Makes the selector added for Navigation fallbacks private.

Closes https://github.com/WordPress/gutenberg/issues/50372

## Why?

Because we're not 100% confident on the API yet.

## How?

1. Move to private selectors
2. Implement the locking and unlocking apis in the `core-data` package
3. Update the `data` package to map private selectors to their resolvers
4. Register the new locked navigation fallback selector

## Testing Instructions

- all tests should pass
- the navigation block fallback mechanism should work
- the site editor navigation sidebar fallback mechanism should work
- the `getNavigationFallbackId` should not accessible without unlocking the core store

### CODE REVIEW

**Comment by getdave** [comment]
@draganescu I started on this but I'm not sure how to get the `unlock` API so it's available in the Navigation block code where the selector is accessed. Core Data doesn't seem to have this setup enabled yet.

**Comment by draganescu** [comment]
Yes it looks like the private selectors are private in the sense they're not exported, not in the sense we use the locking API. Is it difficult to implement lock/unloc in this PR as well? Can I help?

**Comment by getdave** [comment]
Yeh I guess I'm not sure how to go about implementing the lock/unlock stuff for Core Data. Other modules have it. Any help much appreciated.

**Comment by draganescu** [comment]
The problem here is that it looks like locked selectors do not call their resolvers.

**Comment by getdave** [comment]
> The problem here is that it looks like locked selectors do not call thir resolvers.

Ah that's not great. Do you have a code path for that problem?

**Comment by draganescu** [comment]
@getdave the data package in the `createReduxStore` function missed a mapping of the private selectors to resolvers - see https://github.com/WordPress/gutenberg/pull/51413/commits/00844b5d5bf1b27b36ca6ce31e46574f8e2e9743

**Comment by tellthemachines** [comment]
Hey peeps, does this need to go into WP 6.3?

File: packages/data/src/redux-store/index.js [hunk 1]
```diff
@@ -259,6 +259,12 @@ export default function createReduxStore( key, options ) {
 					store,
 					resolversCache
 				);
+				privateSelectors = mapSelectorsWithResolvers(
+					privateSelectors,
+					resolvers,
+					store,
+					resolversCache
+				);
```
Comment (dmsnell) [inline packages/data/src/redux-store/index.js]: great fix, and thanks for following up on this @draganescu

**Comment by getdave** [comment]
Testing now

**Comment by getdave** [comment]
I found a bug with one usage in the Browse Mode sidebar that wasn't unlocked. Other than that this tests well.

@tellthemachines This needs to land in 6.3 🙏

**Review by getdave — COMMENTED** [review]
I found (and fixed) a bug with one usage in the Browse Mode sidebar that wasn't unlocked. Other than that this tests well.

@tellthemachines This needs to land in 6.3 🙏 

@scruffian I cannot approve this cos I raised the PR so can you ✅ ?

File: packages/data/src/redux-store/index.js (hunk 1 above)
Comment (jsnajdr) [reply packages/data/src/redux-store/index.js]: Yes, this is probably the first time we're creating a private selector with a resolver :+1: However, this fix won't work when you register a private selector _after_ a store got instantiated:
```js
const store = createReduxStore( name, { ..., resolvers: { foo } );
register( store ); // instance is created and resolvers bound
unlock( store ).registerPrivateSelectors( { foo } );
```
Here you supplied a resolver for `foo`, registered a private selector `foo`, but they won't be bound together. You'd have to move the `register` call after the private registration.

That's why the private selectors are bound lazily. I'll try to do a more perfect fix in a separate PR.

Comment (draganescu) [reply packages/data/src/redux-store/index.js]: I specifically moved the registration after to get this effect 🤷🏻

Comment (jsnajdr) [reply packages/data/src/redux-store/index.js]: Yes, you had to move the registration, but you shouldn't have to, that's the bug 🙂 

I think this PR is mergeable with the simple resolver fix. The correct solution turns out the be a rather big PR -- although it doesn't change a lot of code, it involves a lot of refactoring.

**Review by scruffian — CHANGES_REQUESTED** [review]
This breaks the navigation section of Site View, if you have no navigation menus.

**Comment by draganescu** [comment]
It seems the update in the `data` package causes a caching problem for locked selectors and their resolvers are always called

The `getNavigationFallbackId` selector and resolver couple behaves like this:

```
This Branch

selector 
resolver 
selector 
resolver 
selector 
selector 
resolver 


Trunk

selector 
resolver 
selector 
selector 
selector 
```

**Comment by jsnajdr** [comment]
> caching problem for locked selectors and their resolvers are always called

I think it's because for private selectors, `bindSelector` and `mapSelectorsWithResolvers` are called in wrong order. Normally, we would call `bindSelector` first, which binds the selector to store `state`, and then wrap the bound selector with `mapSelectorsWithResolvers`. But this PR does it in the opposite order. The `args` array is then wrong, it also contains the `state` parameter. Caching the results, where the cache is keyed by the `args`, then doesn't work.

We could merge #52036 to unblock this. I'm not happy with the PR, but it works.

**Comment by draganescu** [comment]
@jsnajdr can we just fix the problem here as step 1? You say in https://github.com/WordPress/gutenberg/pull/52036 that you'd like to merge that in multiple PRs - this can be the 1st - as long as we can only fix this problem? Just trying my luck here really :D

**Comment by jsnajdr** [comment]
> can we just fix the problem here as step 1?

I can try to do a minimalistic solution right here in this PR. #52036 is bigger than I originally expected especially because I wanted to add a good unit test, and that requires a working `unlock( resolveSelect( store ) ).privateSelector`, and implementing that required many additional changes.

**Comment by jsnajdr** [comment]
Pushed a fix for the private selectors with resolvers. I also had to remove `core-data`'s duplicate lock/unlock registration to make the unit tests pass.

**Review by scruffian — APPROVED** [review]
This is now working well for me.

**Comment by tellthemachines** [comment]
I just cherry-picked this PR to the update/wp6-3-beta3 branch to get it included in the next release: 7b8db97db1

**Pull request merged**